File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    static boolean[][] vistied;

    static int result = -1;

    static int row,col;

    public class Node {
        int x;
        int y;
        int dist;

        Node(int x, int y, int dist) {
          this.x = x;
          this.y = y;
          this.dist = dist;
        }
    }

    public int solution(int[][] maps) {
        int answer = 0;

        row = maps.length;
        col = maps[0].length;

        vistied = new boolean[row][col];

        bfs(0, 0, maps);

        answer = result;

        return answer;
    }

    public void bfs(int a, int b, int[][] maps) {
        Queue<Node> queue = new LinkedList<>();
        queue.add(new Node(a, b, 1));
        vistied[a][b] = true;

        while(!queue.isEmpty()) {
        Node node = queue.poll();

        if(node.x == maps.length -1 && node.y == maps[0].length - 1) {
            result = node.dist;
        }

            for(int i = 0; i < 4; i++) {
                int nextX = node.x + dx[i];
                int nextY = node.y + dy[i];
                int nextDist = node.dist + 1;

                if(nextX < 0 || nextX >= row || nextY < 0 || nextY >= col) continue;

                if(maps[nextX][nextY] == 0) continue;

                if(vistied[nextX][nextY]) continue;

                queue.add(new Node(nextX, nextY, nextDist));
                vistied[nextX][nextY] = true;

            }
        }
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    
    queue = deque()
    n = len(maps)
    m = len(maps[0])
    visited = [[False] * m for _ in range(n)]
    dist_array = [[0] * m for _ in range(n)]
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    queue.append((0, 0, 1))
    visited[0][0] = True
    
    while queue:
        x, y, dist = queue.popleft()
        
        for i in range(4):
            next_x = x + dx[i]
            next_y = y + dy[i]
            next_dist = dist + 1
            
            if next_x < 0 or next_x >= n or next_y < 0 or next_y >= m:
                continue
            
            if visited[next_x][next_y] or maps[next_x][next_y] == 0:
                continue
            
            visited[next_x][next_y] = True
            dist_array[next_x][next_y] = next_dist
            queue.append((next_x, next_y, next_dist))
    
    if dist_array[n-1][m-1] == 0:
        answer = -1
    else:
        answer = dist_array[n-1][m-1]
    
    return answer